### PR TITLE
Added Stored Proc and new UDFs for Stats Lib

### PR DESCRIPTION
**Pull request by jrossthomson**
Some new functions added to the Stats Lib. Also trying out stored procedures. This is not 100% clear how to use, so we made a first pass.
@imathews 
@boaguilar

### CODE REVIEW

File: udfs/community/mannwhitneyu.sqlx [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    CASE alt
+      WHEN 'less' THEN ( n1n2/2.0 +0.5 - U1)/den
+      WHEN 'greater' THEN (n1n2/2.0 +0.5 - U2)/den
+      ELSE    -1.0*ABS(n1n2/2.0 + 0.5 - IF(U1<U2,U2,U1))/den
+    END as z, 
+    U1 as U,
+    IF( alt='less' OR alt='greater', 1.0, 2.0 ) as factor
+  FROM statistics  
+)
+ SELECT struct(U, factor* fn.normal_cdf(z,0.0,1.0) as p )
```
Comment (danieldeleo) [inline udfs/community/mannwhitneyu.sqlx:55]:
```suggestion
 SELECT struct(U, factor* ${ref("normal_cdf")}(z,0.0,1.0) as p ) 
```

File: udfs/community/mannwhitneyu.sqlx [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+      WHEN 'greater' THEN (n1n2/2.0 +0.5 - U2)/den
+      ELSE    -1.0*ABS(n1n2/2.0 + 0.5 - IF(U1<U2,U2,U1))/den
+    END as z, 
+    U1 as U,
+    IF( alt='less' OR alt='greater', 1.0, 2.0 ) as factor
+  FROM statistics  
+)
+ SELECT struct(U, factor* fn.normal_cdf(z,0.0,1.0) as p ) 
+ FROM normal_appr 
+));
```
Comment (danieldeleo) [inline udfs/community/mannwhitneyu.sqlx:57]:
```suggestion
));

```

**Comment by jrossthomson** [comment]
@danieldeleo Hi Dan, does this PR need anything from me?

File: stored_procedures/linear_regression.sql [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+     UNION ALL SELECT 5.9,3.0,5.1,1.8,'virginica';
+
+
+  CALL bqutil.procedure.linear_regression('iris', 'sepal_width', 'petal_width', result);
+
+  -- We round to 11 decimals here because there appears to be some inconsistency in the function, likely due to floating point errors and the order of aggregation
+  ASSERT ROUND(result.a, 11) = 3.11519268710;
+  ASSERT ROUND(result.b, 11) = -0.62754617565;
+  ASSERT ROUND(result.r, 11) = -0.35654408961;
+END;
```
Comment (danieldeleo) [inline stored_procedures/linear_regression.sql:204]: Add trailing newline

File: stored_procedures/bh_multiple_tests.sql [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+       )
+       FROM bh_multiple_tests_results
+   ) = 8;
+
+
+
+
+END;
+
+
```
Comment (danieldeleo) [inline stored_procedures/bh_multiple_tests.sql:76]: Remove extra trailing line

File: stored_procedures/README.md [hunk 5]
```diff
@@ -29,3 +29,64 @@ END;
 
 IDs are: [99, 100, 101, 102, 103, 104, 105, 106, 107, 108, 109]
 ```
+
```
Comment (danieldeleo) [inline stored_procedures/README.md:32]: LInear regression missing from README

File: stored_procedures/chi_square.sql [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+        UNION ALL SELECT 'cat', 'yarn'
+        UNION ALL SELECT 'cat', 'yarn'
+        UNION ALL SELECT 'cat', 'yarn';
+
+  CALL bqutil.procedure.chi_square('categorical', 'animal', 'toy', result);
+
+  ASSERT result.x = 3.7452380952380966;
+  ASSERT result.dof = 1;
+  ASSERT result.p = 0.052958181867438725;
+END;
```
Comment (danieldeleo) [inline stored_procedures/chi_square.sql:64]:
```suggestion
END;

```

File: stored_procedures/linear_regression.sql (hunk 3 above)
Comment (danieldeleo) [inline stored_procedures/linear_regression.sql:204]:
```suggestion
END;

```

File: udfs/community/chisquare_cdf.sqlx [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+#standardSQL
+
+CREATE OR REPLACE FUNCTION ${self()}(H FLOAT64, dof INT64)
+RETURNS FLOAT64 
+LANGUAGE js AS """
+  return 1.0 - jstat.jStat['chisquare'].cdf(H, dof)
+"""
+OPTIONS ( 
+    library=["${JS_BUCKET}/jstat-v1.9.4.min.js"]
+);
```
Comment (danieldeleo) [inline udfs/community/chisquare_cdf.sqlx:26]:
```suggestion
);

```

File: udfs/community/corr_pvalue.sqlx [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+   var t =  abs_r * Math['sqrt']( (n-2) / (1.0 - (r*r)) )
+   return jstat['jStat']['ttest'](t,n-2,2);
+
+} else if (abs_r == 1.0  ) {
+   return 0.0
+} else {  
+   return NaN 
+}
+"""
+OPTIONS (library=["${JS_BUCKET}/jstat-v1.9.4.min.js"]);
```
Comment (danieldeleo) [inline udfs/community/corr_pvalue.sqlx:37]:
```suggestion
OPTIONS (library=["${JS_BUCKET}/jstat-v1.9.4.min.js"]);

```

File: udfs/community/jstat.sqlx [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+    let fn = jstat['jStat']
+    for (const name of methodPath){
+        fn = fn[name]
+    }
+
+  return fn(...args)
+"""
+OPTIONS (
+	    library=["${JS_BUCKET}/jstat-v1.9.4.min.js"]
+);
```
Comment (danieldeleo) [inline udfs/community/jstat.sqlx:32]:
```suggestion
);

```

File: udfs/community/kruskal_wallis.sqlx [hunk 10]
```diff
@@ -33,5 +33,5 @@ CREATE OR REPLACE FUNCTION ${self()}(data ARRAY<STRUCT<factor STRING, val FLOAT6
                       COUNT(n) -1 AS DoF
         FROM sums
     ) # H_raw
-    SELECT struct(H AS H, ${ref("pvalue")}(H, DoF) AS p, DoF AS DoF) FROM H_raw
+    SELECT struct(H AS H, ${ref("chisquare_cdf")}(H, DoF) AS p, DoF AS DoF) FROM H_raw
 ));
```
Comment (danieldeleo) [inline udfs/community/kruskal_wallis.sqlx:37]:
```suggestion
));

```

File: udfs/community/mannwhitneyu.sqlx [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+      WHEN 'greater' THEN (n1n2/2.0 +0.5 - U2)/den
+      ELSE    -1.0*ABS(n1n2/2.0 + 0.5 - IF(U1<U2,U2,U1))/den
+    END as z, 
+    U1 as U,
+    IF( alt='less' OR alt='greater', 1.0, 2.0 ) as factor
+  FROM statistics  
+)
+ SELECT struct(U, factor* ${ref("normal_cdf")}(z,0.0,1.0) as p ) 
+ FROM normal_appr 
+));
```
Comment (danieldeleo) [inline udfs/community/mannwhitneyu.sqlx:57]:
```suggestion
));

```

File: udfs/community/normal_cdf.sqlx [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+ */
+
+CREATE OR REPLACE FUNCTION ${self()}(x FLOAT64, mean FLOAT64, std FLOAT64)
+RETURNS FLOAT64
+LANGUAGE js AS """
+  return jstat['jStat']['normal']['cdf']( x, mean, std )
+"""
+OPTIONS (
+	    library=["${JS_BUCKET}/jstat-v1.9.4.min.js"]
+);
```
Comment (danieldeleo) [inline udfs/community/normal_cdf.sqlx:26]:
```suggestion
);

```

File: udfs/community/p_fisherexact.sqlx [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+        }
+        if ( LnPrx <= LnPra ) { 
+            temp = temp + Math['exp']( LnPrx - LnPra );
+        }
+     }
+  }
+  
+  var LnPFET = LnPra  + Math['log']( temp ); 
+  return  Math['exp']( LnPFET ) ;
+""";
```
Comment (danieldeleo) [inline udfs/community/p_fisherexact.sqlx:132]:
```suggestion
""";

```

File: udfs/community/t_test.sqlx [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+            STDDEV(p2) st2,
+            COUNT(p2) AS n2
+        FROM UNNEST(pop2) AS p2
+    )
+    SELECT
+        STRUCT(
+            ABS(x1 - x2) / Sqrt((st1 * st1 / n1) + (st2 * st2 / n2)) AS t_value,
+            n1 + n2 - 2 AS dof)
+    FROM pop1 CROSS JOIN pop2
+));
```
Comment (danieldeleo) [inline udfs/community/t_test.sqlx:47]:
```suggestion
));

```

File: stored_procedures/bh_multiple_tests.sql (hunk 4 above)
Comment (danieldeleo) [inline stored_procedures/bh_multiple_tests.sql:76]:
```suggestion
END;

```

File: stored_procedures/linear_regression.sql [hunk 15]
```diff
@@ -0,0 +1,204 @@
+-- @param STRING table_name table (or subquery) that contains the data
```
Comment (danieldeleo) [inline stored_procedures/linear_regression.sql:1]: Add license header

File: stored_procedures/chi_square.sql [hunk 16]
```diff
@@ -0,0 +1,65 @@
+-- @param STRING table_name table (or subquery) that contains the data
```
Comment (danieldeleo) [inline stored_procedures/chi_square.sql:1]: Add license header

File: stored_procedures/bh_multiple_tests.sql [hunk 17]
```diff
@@ -0,0 +1,75 @@
+-- Adjust p values using the Benjamini-Hochberg multipletests method, additional details in doi:10.1098/rsta.2009.0127
```
Comment (danieldeleo) [inline stored_procedures/bh_multiple_tests.sql:1]: Add license header

File: stored_procedures/chi_square.sql (hunk 16 above)
Comment (jrossthomson) [reply stored_procedures/chi_square.sql:17]: Completed

File: stored_procedures/README.md [hunk 18]
```diff
@@ -14,7 +14,10 @@ CALL bqutil.procedure.GetNextIds(10, next_ids);
 
 ## Stored Procedures
 
-* [GetNextIds](#GetNextIds)
+* [GetNextIds](#getnextidsid_count-int64-out-next_ids-array)
+* [chi_square](#chi_squaretable_name-string-independent_var-string-dependent_var-string-out-result-structx-float64-dof-float64-p-float64)
+* [Benjamini-Hochberg multipletests](#bh_multiple_tests-pvalue_table_name-string-pvalue_column_name-string-n_rows-int64-temp_table_name-string-)
+* [Linear Regression](#linear_regression-table_name-string-independent_var-string-dependent_var-string-out-result-structa-float64-b-float64-r-float64-)
```
Comment (danieldeleo) [inline stored_procedures/README.md:20]: Last nit, please rename these to the actual udf names
```suggestion
* [bh_multiple_tests](#bh_multiple_tests-pvalue_table_name-string-pvalue_column_name-string-n_rows-int64-temp_table_name-string-)
* [linear_regression](#linear_regression-table_name-string-independent_var-string-dependent_var-string-out-result-structa-float64-b-float64-r-float64-)
```

**Review by danieldeleo — CHANGES_REQUESTED** [review]
Last nit change and then I'll merge, thanks for your patience!

File: stored_procedures/README.md (hunk 18 above)
Comment (jrossthomson) [reply stored_procedures/README.md]: Done!

**Review by danieldeleo — APPROVED** [review]

**Review by danieldeleo — CHANGES_REQUESTED** [review]

File: udfs/community/test_cases.js [hunk 19]
```diff
@@ -902,6 +902,50 @@ generate_udf_test("linear_regression", [
...
+generate_udf_test("mannwhitneyu", [
+    {
+        inputs: [
+		`(SELECT ARRAY<FLOAT64>[2, 4, 6, 2, 3, 7, 5, 1.])`,
+		`(SELECT ARRAY<FLOAT64>[8, 10, 11, 14, 20, 18, 19, 9.])`, 
+		`CAST('two-sided' AS STRING)`
+	],
+	expected_output: `STRUCT(CAST(64.0 AS FLOAT64) AS U, CAST(9.391056991171487E-4 AS FLOAT64) AS p)`
+    },
+]);
```
[the remark this answers is not in the document]
Comment (jrossthomson) [reply udfs/community/test_cases.js:957]: Removing `jstat` from this PR. Added tests and Readme stuff.

@danieldeleo PTAL. Thanks.

**Review by danieldeleo — APPROVED** [review]

**Pull request merged**